File: .claude/hooks/file-changed-docs-consistency.mjs
#!/usr/bin/env node
// @hookstack file-changed-docs-consistency
// Rappelle de propager les changements d'un README vers les surfaces sœurs
// (FileChanged README.md). Quand un README change, liste les autres README du
// repo (racine + packages/*) qui portent la même promesse produit et doivent
// rester cohérents (exemples CLI, slugs, wording).
import { readFileSync, existsSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

// Trouve les README "surfaces produit" : racine + packages/*/README.md.
export function findSiblingReadmes({
  exists = existsSync,
  readdir = readdirSync,
  projectDir,
} = {}) {
  const surfaces = [];
  if (exists(join(projectDir, 'README.md'))) surfaces.push('README.md');
  const pkgsDir = join(projectDir, 'packages');
  if (exists(pkgsDir)) {
    try {
      for (const pkg of readdir(pkgsDir)) {
        if (exists(join(pkgsDir, pkg, 'README.md'))) surfaces.push(`packages/${pkg}/README.md`);
      }
    } catch {}
  }
  return surfaces;
}

export function run(input, {
  exists = existsSync,
  readdir = readdirSync,
  projectDir = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
} = {}) {
  const filePath = input.file_path ?? '';
  if (!filePath.endsWith('README.md') || input.event === 'unlink') return null;

  const changed = filePath.startsWith(`${projectDir}/`)
    ? filePath.slice(projectDir.length + 1)
    : filePath;
  const siblings = findSiblingReadmes({ exists, readdir, projectDir }).filter((s) => s !== changed);
  if (!siblings.length) return null;

  return {
    hookSpecificOutput: {
      hookEventName: 'FileChanged',
      additionalContext:
        `${changed} changed. These sibling docs share the same product promise and must stay consistent ` +
        `(CLI examples, slugs, wording): ${siblings.join(', ')}. ` +
        'Check whether the change needs to be mirrored there (and on the website copy if user-facing).',
    },
  };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/file-changed-docs-consistency.py
#!/usr/bin/env python3
# @hookstack file-changed-docs-consistency
"""Reminds to propagate README changes to sibling doc surfaces (FileChanged)."""
import json
import os
import sys


def _exists(path):
    return os.path.exists(path)


def _readdir(path):
    return os.listdir(path)


def find_sibling_readmes(*, exists=None, readdir=None, project_dir):
    if exists is None:
        exists = _exists
    if readdir is None:
        readdir = _readdir

    surfaces = []
    if exists(os.path.join(project_dir, "README.md")):
        surfaces.append("README.md")
    pkgs_dir = os.path.join(project_dir, "packages")
    if exists(pkgs_dir):
        try:
            for pkg in readdir(pkgs_dir):
                if exists(os.path.join(pkgs_dir, pkg, "README.md")):
                    surfaces.append(f"packages/{pkg}/README.md")
        except Exception:
            pass
    return surfaces


def run(input_data, *, exists=None, readdir=None, project_dir=None):
    if exists is None:
        exists = _exists
    if readdir is None:
        readdir = _readdir
    if project_dir is None:
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())

    file_path = input_data.get("file_path") or ""
    if not file_path.endswith("README.md") or input_data.get("event") == "unlink":
        return None

    changed = (
        file_path[len(project_dir) + 1 :]
        if file_path.startswith(f"{project_dir}/")
        else file_path
    )
    siblings = [
        s
        for s in find_sibling_readmes(exists=exists, readdir=readdir, project_dir=project_dir)
        if s != changed
    ]
    if not siblings:
        return None

    return {
        "hookSpecificOutput": {
            "hookEventName": "FileChanged",
            "additionalContext": (
                f"{changed} changed. These sibling docs share the same product promise and must stay consistent "
                f"(CLI examples, slugs, wording): {', '.join(siblings)}. "
                "Check whether the change needs to be mirrored there (and on the website copy if user-facing)."
            ),
        }
    }


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
